<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>PICtR: Physically Interacting Cell toolkit for R • PICtR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="PICtR: Physically Interacting Cell toolkit for R">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">PICtR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/PICtR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/agSHaas/PICtR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>PICtR: Physically Interacting Cell toolkit for R</h1>
                        <h4 data-toc-skip class="author">Viktoria
Flore</h4>
            
            <h4 data-toc-skip class="date">last updated 01/2025</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/agSHaas/PICtR/blob/HEAD/vignettes/PICtR.Rmd" class="external-link"><code>vignettes/PICtR.Rmd</code></a></small>
      <div class="d-none name"><code>PICtR.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>Many fundamental processes in life are shaped by physical
interactions between cells. PICtR provides a computational workflow to
analyse both cellular landscapes and cellular interactions in
high-dimensional flow cytometry data using R. PICtR can be used with any
cytometry-based assay and the approach can also be applied to existing
datasets, provided that the following guidelines are met:</p>
<ul>
<li>The data stems from a case-control setting</li>
<li>All samples were treated the same during sample processing and
acquisition, e.g. regarding equal cell concentrations, incubation times,
and flow rates</li>
<li>Pooling across conditions, organs or donors should be avoided</li>
<li>For <em>in vivo</em> experiments, cells with a strong affinity to
interact can artificially interact if they are brought into proximity
during sample preparation, even if they were physically separated <em>in
vivo</em>. These interactions may still reflect biological effects, but
this should be taken into account when interpreting the results.</li>
</ul>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>PICtR requires R version 4.3 or later. First, please install <a href="https://github.com/bnprks/BPCells" class="external-link">BPCells</a>. Next, PICtR can be
installed using:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://remotes.r-lib.org" class="external-link">"remotes"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"remotes"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="fu">remotes</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"agSHaas/PICtR"</span><span class="op">)</span></span></code></pre></div>
<p>Some other packages are not necessarily required, but expand the
functionality of PICtR. Consider installing:</p>
<ul>
<li>
<a href="https://www.bioconductor.org/packages/release/bioc/html/flowMeans.html" class="external-link">flowMeans</a><br>
</li>
<li>
<a href="https://github.com/mhahsler/dbscan" class="external-link">dbscan</a><br>
</li>
<li><a href="https://github.com/ImmuneDynamics/Spectre" class="external-link">Spectre</a></li>
<li>
<a href="https://www.bioconductor.org/packages/release/bioc/html/ComplexHeatmap.html" class="external-link">ComplexHeatmap</a><br>
</li>
<li>
<a href="https://www.bioconductor.org/packages/release/bioc/html/cytoMEM.html" class="external-link">cytoMEM</a><br>
</li>
<li><a href="https://github.com/rorynolan/autothresholdr" class="external-link">autothresholdr</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="pictr-pipeline">PICtR Pipeline<a class="anchor" aria-label="anchor" href="#pictr-pipeline"></a>
</h2>
<div class="section level3">
<h3 id="pre-processing-of-cytometry-data">Pre-Processing of Cytometry Data<a class="anchor" aria-label="anchor" href="#pre-processing-of-cytometry-data"></a>
</h3>
<p>Data from FCS files should be unmixed/compensated and transformed
(for example using FlowJo from BD). We recommend using quality control
tools such as <a href="https://doi.org/10.1002/cyto.a.24501" class="external-link">PeacoQC</a>
or <a href="https://doi.org/10.1093/bioinformatics/btw191" class="external-link">FlowAI</a> to
ensure high quality cytometry data. Export the populations of interest
using channel values defined by the inbuilt export function of FlowJo or
use <a href="https://github.com/RGLab/flowWorkspace" class="external-link">flowWorkspace</a>.</p>
</div>
<div class="section level3">
<h3 id="example-data">Example Data<a class="anchor" aria-label="anchor" href="#example-data"></a>
</h3>
<p>We provide a small demo data set from the spleens of lymphocytic
choriomeningitis virus (LCMV) infected mice and controls. Five days
prior to infection, LCMV-specific CD4 (CD90_1-positive) T cells were
transferred into the hosts. The samples were analysed using high
parametric spectral flow cytometry with <em>n</em> = 36 markers,
allowing the detection of all important cell types across the immune
cell landscape. For the demo data set, we will focus on 10 different
markers:</p>
<ul>
<li>CD3: T cell marker<br>
</li>
<li>MHCII: Major histocompatibility complex II, expressed on
antigen-presenting cells<br>
</li>
<li>CD11c: Dendritic cell (DC) marker<br>
</li>
<li>CD11b: Myeloid cell marker</li>
<li>CD45_2: Pan-hematopoetic marker, allelic form 2<br>
</li>
<li>CD19: B cell marker<br>
</li>
<li>Ly6G: Neutrophil marker<br>
</li>
<li>CD90_1: Marker for LCMV-specific CD4 T cells<br>
</li>
<li>CD4: Expressed on CD4 T cells and some myeloid populations<br>
</li>
<li>CD8: Expressed on CD8 T cells, some natural killer cells (NK) and
some dendritic cells</li>
</ul>
<p>Details on the experimental setup can be found in Vonficht D,
Jopp-Saile L, Yousefian S, Flore V et al. Ultra-high scale
cytometry-based cellular interaction mapping. <em>Nature Methods</em>
(2025).</p>
<p>Demo data can be loaded with <code>data(demo_lcmv)</code>. See
<code><a href="../reference/demo_lcmv.html">?demo_lcmv</a></code> for details.</p>
</div>
<div class="section level3">
<h3 id="subsampling-dimensional-reduction-and-clustering">Subsampling, Dimensional Reduction and Clustering<a class="anchor" aria-label="anchor" href="#subsampling-dimensional-reduction-and-clustering"></a>
</h3>
<p>The pipeline starts with a data frame with the dimensionality cells x
flow cytometry parameters. First, we calculate the FSC ratio based on
the forward scatter area (FSC-A) and forward scatter height (FSC-H)
parameters. The FSC ratio is highly discriminatory for singlets and
multiplets, and will help us to select interacting populations after
clustering. We find a threshold that classifies all events as having a
low or high FSC ratio using thresholding methods such as Otsu
thresholding.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://agshaas.github.io/PICtR/">PICtR</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"demo_lcmv"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">threshold</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/calculateThreshold.html">calculateThreshold</a></span><span class="op">(</span><span class="op">(</span><span class="va">demo_lcmv</span><span class="op">$</span><span class="va">FSC.A</span><span class="op">/</span><span class="va">demo_lcmv</span><span class="op">$</span><span class="va">FSC.H</span><span class="op">)</span>, method <span class="op">=</span> <span class="st">"otsu"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/hist.html" class="external-link">hist</a></span><span class="op">(</span><span class="va">demo_lcmv</span><span class="op">$</span><span class="va">FSC.A</span><span class="op">/</span><span class="va">demo_lcmv</span><span class="op">$</span><span class="va">FSC.H</span>, breaks <span class="op">=</span> <span class="fl">2000</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/abline.html" class="external-link">abline</a></span><span class="op">(</span>v <span class="op">=</span> <span class="va">threshold</span><span class="op">)</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/FSC_ratio-1.png" width="672"></p>
<p>Then, we create a <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>
object using forward scatter parameters, side scatter parameters, cell
type markers and the FSC ratio as features. Several steps of the PICtR
workflow make use of the Seurat framework.<br>
We store the counts layer on-disk using <a href="https://github.com/bnprks/BPCells" class="external-link">BPCells</a>, enabling us to
analyse millions of events while keeping computational time manageable.
Thus far, we have analysed data sets with more than 60 million events
with PICtR.<br>
In addition, we sample a representative subset of cells using atomic
sketching as implemented in <a href="https://satijalab.org/seurat/articles/parsebio_sketch_integration" class="external-link">Seurat</a>.
In contrast to random sampling, sketching retains both abundant and rare
cell types, including cellular interactions. For the demo data set, we
sample <code>n_sketch_cells = 5000</code> cells. We use the sketched
data set for dimensional reduction and clustering, before predicting
cell type and interaction labels back to the full data set later on.</p>
<p>PICtR provides a convenient wrapper function for these steps,
including calculation of the FSC ratio threshold:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sketch_wrapper.html">sketch_wrapper</a></span><span class="op">(</span><span class="va">demo_lcmv</span>,</span>
<span>                       meta_data <span class="op">=</span> <span class="va">demo_lcmv</span>,</span>
<span>                       assay <span class="op">=</span> <span class="st">"FACS"</span>,</span>
<span>                       FSC.A <span class="op">=</span> <span class="st">"FSC.A"</span>, </span>
<span>                       FSC.H <span class="op">=</span> <span class="st">"FSC.H"</span>, </span>
<span>                       working_dir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>                       resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.8</span>, <span class="fl">1</span><span class="op">)</span>,</span>
<span>                       clst_algorithm <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                       n_sketch_cells <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>                       obj_name <span class="op">=</span> <span class="st">"obj_sketched_non_projected"</span>,</span>
<span>                       ratio <span class="op">=</span> <span class="cn">T</span>, </span>
<span>                       thresholding_method <span class="op">=</span> <span class="st">"otsu"</span>, </span>
<span>                       overwrite <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt; The BPCells directory will be overwritten</span></span>
<span><span class="co">#&gt; Calculation of Ratio</span></span>
<span><span class="co">#&gt; Warning: Matrix compression performs poorly with non-integers.</span></span>
<span><span class="co">#&gt; <span style="color: #00BBBB;">•</span> Consider calling convert_matrix_type if a compressed integer matrix is intended.</span></span>
<span><span class="co">#&gt; Your newly generated object will be saved under: /Users/nflore/Documents/01_PICtR/PICtR/vignettes/obj_sketched_non_projected.rds</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Sketching started...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Finding variable features for layer counts</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Calcuating Leverage Score</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attempting to cast layer counts to dgCMatrix</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Attempting to cast layer data to dgCMatrix</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Sketching is done</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Finding variable features for layer counts</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Centering and scaling data matrix</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; PC_ 1 </span></span>
<span><span class="co">#&gt; Positive:  CD4, Ly6G, CD11b, CD90-1, CD8, CD11c, SSC.H, MHCII </span></span>
<span><span class="co">#&gt; Negative:  FSC.A, ratio, SSC.A, FSC.H, CD45-2, CD3, CD19 </span></span>
<span><span class="co">#&gt; PC_ 2 </span></span>
<span><span class="co">#&gt; Positive:  CD4, CD45-2, CD3, CD90-1, CD8, MHCII, CD19, ratio </span></span>
<span><span class="co">#&gt; Negative:  CD11b, SSC.H, Ly6G, SSC.A, FSC.H, CD11c, FSC.A </span></span>
<span><span class="co">#&gt; PC_ 3 </span></span>
<span><span class="co">#&gt; Positive:  CD4, CD90-1, CD3, SSC.H, CD11b, SSC.A, FSC.H, CD8 </span></span>
<span><span class="co">#&gt; Negative:  MHCII, CD19, ratio, CD45-2, FSC.A, CD11c, Ly6G </span></span>
<span><span class="co">#&gt; PC_ 4 </span></span>
<span><span class="co">#&gt; Positive:  CD19, CD90-1, CD4, Ly6G, ratio, FSC.A, SSC.A, MHCII </span></span>
<span><span class="co">#&gt; Negative:  CD11c, CD8, CD45-2, CD11b, CD3, SSC.H, FSC.H </span></span>
<span><span class="co">#&gt; PC_ 5 </span></span>
<span><span class="co">#&gt; Positive:  CD11c, MHCII, CD4, CD90-1, CD45-2, SSC.H, CD11b, SSC.A </span></span>
<span><span class="co">#&gt; Negative:  CD8, Ly6G, CD19, CD3, FSC.H, FSC.A, ratio </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This message is displayed once every 8 hours.</span></span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 5000</span></span>
<span><span class="co">#&gt; Number of edges: 164643</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.9193</span></span>
<span><span class="co">#&gt; Number of communities: 14</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 5000</span></span>
<span><span class="co">#&gt; Number of edges: 164643</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8959</span></span>
<span><span class="co">#&gt; Number of communities: 16</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 5000</span></span>
<span><span class="co">#&gt; Number of edges: 164643</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8813</span></span>
<span><span class="co">#&gt; Number of communities: 19</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span>
<span><span class="co">#&gt; UMAP will return its model</span></span>
<span><span class="co">#&gt; 14:55:28 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 14:55:28 Read 5000 rows and found 15 numeric columns</span></span>
<span><span class="co">#&gt; 14:55:28 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 14:55:28 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 14:55:28 Writing NN index file to temp file /var/folders/zq/k957_k5s79v97_ptgc729fm1yy94mx/T//RtmpC2SNKZ/file64fb61fc9668</span></span>
<span><span class="co">#&gt; 14:55:28 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 14:55:29 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 14:55:29 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 14:55:29 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">#&gt; 14:55:29 Commencing optimization for 500 epochs, with 202814 positive edges</span></span>
<span><span class="co">#&gt; 14:55:29 Using rng type: pcg</span></span>
<span><span class="co">#&gt; 14:55:35 Optimization finished</span></span>
<span><span class="co">#&gt; The object will be updated and saved</span></span></code></pre></div>
<div class="section level4">
<h4 id="guidelines-for-clustering">Guidelines for Clustering<a class="anchor" aria-label="anchor" href="#guidelines-for-clustering"></a>
</h4>
<p>Finding a good clustering resolution is critical to detecting and
annotating interacting populations. PICtR provides several different
options for clustering: In addition to <a href="10.1088/1742-5468/2008/10/P10008">Louvain</a>, <a href="https://doi.org/10.1140/epjb/e2013-40829-0" class="external-link">SLM</a>, and <a href="https://doi.org/10.1038/s41598-019-41695-z" class="external-link">Leiden</a> algorithms,
which are implemented in the Seurat framework, <a href="https://doi.org/10.1002/cyto.a.22625" class="external-link">FlowSOM</a>, <a href="https://doi.org/10.1002/cyto.a.21007" class="external-link">flowMeans</a> and <a href="https://doi.org/10.1007/978-3-642-37456-2_14" class="external-link">HDBSCAN</a>
(Hierarchical Density-Based Spatial Clustering of Applications with
Noise) are available. We recommend using Louvain or Leiden due to their
successful identification of interacting populations, fast run times,
and easy customization of the resolution parameter (note that Leiden
requires the leidenalg python package).</p>
<p>The effect of the resolution parameter can be seen when comparing a
resolution of 0.5, 0.8 and 1 (Louvain clustering) for the demo data
set:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">demo_obj</span>, group.by <span class="op">=</span> <span class="st">"sketch_snn_res.0.5"</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">demo_obj</span>, group.by <span class="op">=</span> <span class="st">"sketch_snn_res.0.8"</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">demo_obj</span>, group.by <span class="op">=</span> <span class="st">"sketch_snn_res.1"</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/clustering_resolution-1.png" width="700"><img src="PICtR_files/figure-html/clustering_resolution-2.png" width="700"><img src="PICtR_files/figure-html/clustering_resolution-3.png" width="700"></p>
<p>We recommend to over-cluster the data to ensure the detection of
small populations, including cellular interactions. So if you expect
roughly 20 different cell types based on your flow cytometry panel, aim
for a cluster resolution that provides you with more than 20 clusters.
Subsequently, during annotation, clusters representing the same cell
type can be merged. Alternatively, for clusters that represent
populations of interest that require higher resolutions, sub-clustering
can be performed (<code><a href="https://satijalab.org/seurat/reference/FindSubCluster.html" class="external-link">FindSubCluster()</a></code>). Different resolutions
might be tested before deciding for the best fit.</p>
<p>It can also be helpful to compare your cluster boundaries to feature
plots of the markers used for annotation to see if you capture all
relevant patterns. For our demo data set, a resolution of 0.5 seems to
be sufficient:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">demo_obj</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3"</span>, <span class="st">"MHCII"</span>, <span class="st">"CD11c"</span>, <span class="st">"CD11b"</span>, </span>
<span>                                   <span class="st">"CD45_2"</span>, <span class="st">"CD19"</span>, <span class="st">"Ly6G"</span>, </span>
<span>                                   <span class="st">"CD90_1"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"ratio"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/feature_plots_fig-1.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig-2.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig-3.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig-4.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig-5.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig-6.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig-7.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig-8.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig-9.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig-10.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig-11.png" width="192"></p>
<p>The <a href="https://satijalab.org/seurat/" class="external-link">Seurat tutorials</a> and
<a href="https://doi.org/10.1038/s41576-023-00586-w" class="external-link">Heumos et
al. (2023)</a> provide excellent guidance on clustering of single-cell
data sets.</p>
</div>
</div>
<div class="section level3">
<h3 id="predict-labels-for-the-full-data-set">Predict Labels for the full Data Set<a class="anchor" aria-label="anchor" href="#predict-labels-for-the-full-data-set"></a>
</h3>
<p>Currently, our cluster labels only exist for the sketched subset of
cells, but we can predict the labels for the entire data set using
linear discriminant analysis (LDA) as implemented in the
<code><a href="../reference/predict_data.html">predict_data()</a></code> function:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/predict_data.html">predict_data</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">demo_obj</span>, </span>
<span>                      data_query <span class="op">=</span> <span class="va">demo_obj</span>, </span>
<span>                      ref_clusters <span class="op">=</span> <span class="st">"sketch_snn_res.0.5"</span>,</span>
<span>                      pred_name <span class="op">=</span> <span class="st">"clusters_predicted"</span>,</span>
<span>                      assay_ref <span class="op">=</span> <span class="st">"sketch"</span>, </span>
<span>                      assay_query <span class="op">=</span> <span class="st">"FACS"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Converting to a dense matrix may use excessive memory</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This message is displayed once every 8 hours.</span></span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="selecting-interacting-populations">Selecting Interacting Populations<a class="anchor" aria-label="anchor" href="#selecting-interacting-populations"></a>
</h3>
<p>Now we can select the clusters that represent interacting cells. For
each cluster, we determine the proportion of cells above and below the
FSC ratio threshold that we calculated earlier:</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/ratio_cluster_plot.html">ratio_cluster_plot</a></span><span class="op">(</span><span class="va">demo_obj</span>, clusters <span class="op">=</span> <span class="st">"clusters_predicted"</span><span class="op">)</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/ratio_cluster_plot-1.png" width="672"></p>
<p>As seen from the bar plot, clusters 3, 7, 10, 11, and 13
predominantly contain cells that exceed the FSC ratio threshold. The
<code><a href="../reference/select_dbt.html">select_dbt()</a></code> function can be used to pick these clusters.
We recommend investigating the <code><a href="../reference/ratio_cluster_plot.html">ratio_cluster_plot()</a></code> to
confirm any selection you make.</p>
</div>
<div class="section level3">
<h3 id="singlet-annotation">Singlet Annotation<a class="anchor" aria-label="anchor" href="#singlet-annotation"></a>
</h3>
<p>Based on the previous analysis, clusters 3, 7, 10, 11 and 13 have
been identified as representing interacting cells. Now, we can annotate
the remaining clusters with the help of the feature plots (see above) or
ridge plots:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">demo_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"clusters_predicted"</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/RidgePlot.html" class="external-link">RidgePlot</a></span><span class="op">(</span><span class="va">demo_obj</span>, features <span class="op">=</span> <span class="fu">Features</span><span class="op">(</span><span class="va">demo_obj</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 23.7</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 20.3</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 9.11</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 15.5</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 24.9</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 21.7</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 26.5</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 32.7</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 28.4</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 43</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 28.9</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 11.3</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 19.9</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 9.44</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 19.1</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/ridge%20plots-1.png" width="672"></p>
<p>Based on the marker expression, we can annotate all clusters as
follows:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">demo_obj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="va">demo_obj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>celltype <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"0"</span> <span class="op">~</span> <span class="st">"B cells"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"1"</span> <span class="op">~</span> <span class="st">"CD8 T cells"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"2"</span> <span class="op">~</span> <span class="st">"DCs"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"3"</span> <span class="op">~</span> <span class="st">"interacting cells"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"4"</span> <span class="op">~</span> <span class="st">"LCMV-spec CD4 T cells"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"5"</span> <span class="op">~</span> <span class="st">"Myeloid cells"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"6"</span> <span class="op">~</span> <span class="st">"CD4 T cells"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"7"</span> <span class="op">~</span> <span class="st">"interacting cells"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"8"</span> <span class="op">~</span> <span class="st">"Myeloid cells"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"9"</span> <span class="op">~</span> <span class="st">"Neutrophils"</span>,</span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"10"</span> <span class="op">~</span> <span class="st">"interacting cells"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"11"</span> <span class="op">~</span> <span class="st">"interacting cells"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"12"</span> <span class="op">~</span> <span class="st">"DCs"</span>, </span>
<span>                              <span class="va">clusters_predicted</span> <span class="op">==</span> <span class="st">"13"</span> <span class="op">~</span> <span class="st">"interacting cells"</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">demo_obj</span>, group.by <span class="op">=</span> <span class="st">"celltype"</span>, label <span class="op">=</span><span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/annotation-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="annotation-of-interacting-populations">Annotation of Interacting Populations<a class="anchor" aria-label="anchor" href="#annotation-of-interacting-populations"></a>
</h3>
<p>Annotating the interacting cells is easier if we subset and
re-cluster them:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># subset interacting cells</span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">demo_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"FACS"</span></span>
<span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">demo_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"celltype"</span></span>
<span><span class="va">interact_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">demo_obj</span>, idents <span class="op">=</span> <span class="st">"interacting cells"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># repeat the Seurat workflow </span></span>
<span><span class="va">n_dims</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu">Features</span><span class="op">(</span><span class="va">demo_obj</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">interact_obj</span> <span class="op">&lt;-</span> <span class="va">interact_obj</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span>npcs<span class="op">=</span><span class="va">n_dims</span>, approx<span class="op">=</span><span class="cn">F</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_dims</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span>resolution <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.5</span>, <span class="fl">0.8</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_dims</span>, return.model <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt; Finding variable features for layer counts</span></span>
<span><span class="co">#&gt; PC_ 1 </span></span>
<span><span class="co">#&gt; Positive:  MHCII, CD19, CD45-2, CD4, Ly6G, CD8, CD90-1, CD11b </span></span>
<span><span class="co">#&gt; Negative:  SSC.A, FSC.A, SSC.H, FSC.H, CD3, CD11c, ratio </span></span>
<span><span class="co">#&gt; PC_ 2 </span></span>
<span><span class="co">#&gt; Positive:  CD4, CD3, CD8, CD11c, CD90-1, CD11b, FSC.H, CD45-2 </span></span>
<span><span class="co">#&gt; Negative:  CD19, MHCII, SSC.A, ratio, FSC.A, SSC.H, Ly6G </span></span>
<span><span class="co">#&gt; PC_ 3 </span></span>
<span><span class="co">#&gt; Positive:  CD45-2, CD4, CD3, ratio, CD19, FSC.A, CD90-1, MHCII </span></span>
<span><span class="co">#&gt; Negative:  Ly6G, CD11b, SSC.H, SSC.A, CD11c, FSC.H, CD8 </span></span>
<span><span class="co">#&gt; PC_ 4 </span></span>
<span><span class="co">#&gt; Positive:  CD90-1, FSC.H, SSC.H, CD4, FSC.A, SSC.A, CD19, MHCII </span></span>
<span><span class="co">#&gt; Negative:  CD45-2, CD11b, CD11c, CD8, CD3, ratio, Ly6G </span></span>
<span><span class="co">#&gt; PC_ 5 </span></span>
<span><span class="co">#&gt; Positive:  CD8, FSC.A, CD11c, FSC.H, ratio, SSC.A, CD19, MHCII </span></span>
<span><span class="co">#&gt; Negative:  CD4, CD45-2, Ly6G, CD90-1, CD11b, CD3, SSC.H</span></span>
<span><span class="co">#&gt; Computing nearest neighbor graph</span></span>
<span><span class="co">#&gt; Computing SNN</span></span>
<span><span class="co">#&gt; UMAP will return its model</span></span>
<span><span class="co">#&gt; 14:55:53 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 14:55:53 Read 9808 rows and found 15 numeric columns</span></span>
<span><span class="co">#&gt; 14:55:53 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 14:55:53 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 14:55:53 Writing NN index file to temp file /var/folders/zq/k957_k5s79v97_ptgc729fm1yy94mx/T//RtmpC2SNKZ/file64fb33177fff</span></span>
<span><span class="co">#&gt; 14:55:53 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 14:55:56 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 14:55:56 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 14:55:57 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">#&gt; 14:55:57 Commencing optimization for 500 epochs, with 406056 positive edges</span></span>
<span><span class="co">#&gt; 14:55:57 Using rng type: pcg</span></span>
<span><span class="co">#&gt; 14:56:06 Optimization finished</span></span>
<span></span>
<span><span class="co"># plot the different clustering resolutions</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">interact_obj</span>, group.by <span class="op">=</span> <span class="st">"FACS_snn_res.0.5"</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">interact_obj</span>, group.by <span class="op">=</span> <span class="st">"FACS_snn_res.0.8"</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">interact_obj</span>, group.by <span class="op">=</span> <span class="st">"FACS_snn_res.1"</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span> <span class="op">+</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.position <span class="op">=</span> <span class="st">"none"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 9808</span></span>
<span><span class="co">#&gt; Number of edges: 329874</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8902</span></span>
<span><span class="co">#&gt; Number of communities: 11</span></span>
<span><span class="co">#&gt; Elapsed time: 1 seconds</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 9808</span></span>
<span><span class="co">#&gt; Number of edges: 329874</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8573</span></span>
<span><span class="co">#&gt; Number of communities: 15</span></span>
<span><span class="co">#&gt; Elapsed time: 1 seconds</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 9808</span></span>
<span><span class="co">#&gt; Number of edges: 329874</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8421</span></span>
<span><span class="co">#&gt; Number of communities: 20</span></span>
<span><span class="co">#&gt; Elapsed time: 1 seconds</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/interacting_analysis-1.png" width="700"><img src="PICtR_files/figure-html/interacting_analysis-2.png" width="700"><img src="PICtR_files/figure-html/interacting_analysis-3.png" width="700"></p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">interact_obj</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3"</span>, <span class="st">"MHCII"</span>, <span class="st">"CD11c"</span>, <span class="st">"CD11b"</span>, </span>
<span>                                   <span class="st">"CD45_2"</span>, <span class="st">"CD19"</span>, <span class="st">"Ly6G"</span>, </span>
<span>                                   <span class="st">"CD90_1"</span>, <span class="st">"CD4"</span>, <span class="st">"CD8"</span>, <span class="st">"ratio"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/feature_plots_fig_interacting-1.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig_interacting-2.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig_interacting-3.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig_interacting-4.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig_interacting-5.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig_interacting-6.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig_interacting-7.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig_interacting-8.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig_interacting-9.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig_interacting-10.png" width="192"><img src="PICtR_files/figure-html/feature_plots_fig_interacting-11.png" width="192"></p>
<p>We can move forward with a resolution of 0.5, but there is a pattern
of CD8 within cluster 4 that is not well represented in this clustering
resolution. We can sub-cluster this population on the same
shared-nearest-neighbor graph (<code>FACS_snn</code>) using
<code><a href="https://satijalab.org/seurat/reference/FindSubCluster.html" class="external-link">FindSubCluster()</a></code> with a low resolution:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">interact_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"FACS_snn_res.0.5"</span></span>
<span><span class="va">interact_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindSubCluster.html" class="external-link">FindSubCluster</a></span><span class="op">(</span><span class="va">interact_obj</span>, cluster <span class="op">=</span> <span class="st">"4"</span>, </span>
<span>                               graph.name <span class="op">=</span> <span class="st">"FACS_snn"</span>, resolution <span class="op">=</span> <span class="fl">0.25</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># plot </span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">interact_obj</span>, group.by <span class="op">=</span> <span class="st">"sub.cluster"</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt; Modularity Optimizer version 1.3.0 by Ludo Waltman and Nees Jan van Eck</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Number of nodes: 845</span></span>
<span><span class="co">#&gt; Number of edges: 29301</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Running Louvain algorithm...</span></span>
<span><span class="co">#&gt; Maximum modularity in 10 random starts: 0.8506</span></span>
<span><span class="co">#&gt; Number of communities: 3</span></span>
<span><span class="co">#&gt; Elapsed time: 0 seconds</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/subcluster-1.png" width="700"></p>
<p>Now we can annotate the interacting cells based on mutually exclusive
marker expression. CD19 for example is only expressed on B cells, while
CD3 and CD4 expression is found on CD4+ T cells. When CD19, CD3 and CD4
are simultaneously expressed within an interacting cluster, we know that
the interaction partners are B cells and CD4+ T cells. We will label
interactions with an asterisk (*):</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://satijalab.github.io/seurat-object/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">interact_obj</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"sub.cluster"</span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/RidgePlot.html" class="external-link">RidgePlot</a></span><span class="op">(</span><span class="va">interact_obj</span>, features <span class="op">=</span> <span class="fu">Features</span><span class="op">(</span><span class="va">interact_obj</span><span class="op">)</span><span class="op">)</span> <span class="op">&amp;</span> </span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">7</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 18.7</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 19.4</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 7.69</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 16.2</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 21.4</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 15.8</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 24</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 30.3</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 23.8</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 47.8</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 28</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 9.23</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 14.4</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 6.68</span></span>
<span><span class="co">#&gt; Picking joint bandwidth of 21.2</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/ridge_plots_interacting-1.png" width="672"></p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">interact_obj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op">&lt;-</span> <span class="va">interact_obj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>interact_type <span class="op">=</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/case_when.html" class="external-link">case_when</a></span><span class="op">(</span><span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"0"</span> <span class="op">~</span> <span class="st">"B cells only"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"1"</span> <span class="op">~</span> <span class="st">"B cells only"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"2"</span> <span class="op">~</span> <span class="st">"B*DC"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"3"</span> <span class="op">~</span> <span class="st">"B*CD4 T"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"4_0"</span> <span class="op">~</span> <span class="st">"DC*LCMV-spec CD4 T"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"4_1"</span> <span class="op">~</span> <span class="st">"B*LCMV-spec CD4 T"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"4_2"</span> <span class="op">~</span> <span class="st">"B*DC*LCMV-spec CD4 T"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"5"</span> <span class="op">~</span> <span class="st">"CD4 T*DC"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"6"</span> <span class="op">~</span> <span class="st">"B*CD8 T"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"7"</span> <span class="op">~</span> <span class="st">"B*Neutrophils"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"8"</span> <span class="op">~</span> <span class="st">"CD4*CD8"</span>, </span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"9"</span> <span class="op">~</span> <span class="st">"B*Myeloid"</span>,</span>
<span>                              <span class="va">sub.cluster</span> <span class="op">==</span> <span class="st">"10"</span> <span class="op">~</span> <span class="st">"remove"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Cluster 10 has a lower CD45_2 expression and a low ratio, so we will
remove it from the interacting landscape. Clusters 0 and 1 do not
exhibit mutually exclusive marker expression - both populations only
express B cell markers. It is likely that they represent homotypic B
cell interactions, but we cannot rule out alternative explanations such
as preceding cytokinesis or interactions with cells that we do not have
a marker for. So we will exclude these populations and recalculate the
UMAP embedding:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">keep</span> <span class="op">&lt;-</span> <span class="va">interact_obj</span><span class="op">@</span><span class="va">meta.data</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">interact_type</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"B*Neutrophils"</span>, </span>
<span>                                     <span class="st">"CD4*CD8"</span>, </span>
<span>                                     <span class="st">"B*CD4 T"</span>,             </span>
<span>                                     <span class="st">"B*Myeloid"</span>,</span>
<span>                                     <span class="st">"B*DC"</span>,</span>
<span>                                     <span class="st">"DC*LCMV-spec CD4 T"</span>,</span>
<span>                                     <span class="st">"B*LCMV-spec CD4 T"</span>,</span>
<span>                                     <span class="st">"B*DC*LCMV-spec CD4 T"</span>,</span>
<span>                                     <span class="st">"B*CD8 T"</span>,</span>
<span>                                     <span class="st">"CD4 T*DC"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">interact_obj</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">interact_obj</span>, cells <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">keep</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">interact_obj</span> <span class="op">&lt;-</span> <span class="va">interact_obj</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> </span>
<span>  <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="va">n_dims</span>, return.model <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="co">#&gt; UMAP will return its model</span></span>
<span><span class="co">#&gt; 14:56:14 UMAP embedding parameters a = 0.9922 b = 1.112</span></span>
<span><span class="co">#&gt; 14:56:14 Read 5482 rows and found 15 numeric columns</span></span>
<span><span class="co">#&gt; 14:56:14 Using Annoy for neighbor search, n_neighbors = 30</span></span>
<span><span class="co">#&gt; 14:56:14 Building Annoy index with metric = cosine, n_trees = 50</span></span>
<span><span class="co">#&gt; 0%   10   20   30   40   50   60   70   80   90   100%</span></span>
<span><span class="co">#&gt; [----|----|----|----|----|----|----|----|----|----|</span></span>
<span><span class="co">#&gt; **************************************************|</span></span>
<span><span class="co">#&gt; 14:56:14 Writing NN index file to temp file /var/folders/zq/k957_k5s79v97_ptgc729fm1yy94mx/T//RtmpC2SNKZ/file64fb329653b6</span></span>
<span><span class="co">#&gt; 14:56:14 Searching Annoy index using 1 thread, search_k = 3000</span></span>
<span><span class="co">#&gt; 14:56:16 Annoy recall = 100%</span></span>
<span><span class="co">#&gt; 14:56:16 Commencing smooth kNN distance calibration using 1 thread with target n_neighbors = 30</span></span>
<span><span class="co">#&gt; 14:56:16 Initializing from normalized Laplacian + noise (using RSpectra)</span></span>
<span><span class="co">#&gt; 14:56:17 Commencing optimization for 500 epochs, with 222518 positive edges</span></span>
<span><span class="co">#&gt; 14:56:17 Using rng type: pcg</span></span>
<span><span class="co">#&gt; 14:56:22 Optimization finished</span></span>
<span></span>
<span><span class="co"># plot </span></span>
<span><span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">interact_obj</span>, group.by <span class="op">=</span> <span class="st">"interact_type"</span>, label <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p><img src="PICtR_files/figure-html/interacting_umap-1.png" width="672"></p>
</div>
<div class="section level3">
<h3 id="downstream-analysis">Downstream Analysis<a class="anchor" aria-label="anchor" href="#downstream-analysis"></a>
</h3>
<p>After annotating the singlet and interacting landscapes, you can
continue with downstream analyses according to your experimental setup.
Often, this will involve the quantification of interactions in the case
and control conditions to test for significant differences.</p>
</div>
</div>
<div class="section level2">
<h2 id="citing-pictr">Citing PICtR<a class="anchor" aria-label="anchor" href="#citing-pictr"></a>
</h2>
<p>If you are using PICtR in your research, please consider citing:</p>
<ul>
<li>Vonficht D, Jopp-Saile L, Yousefian S, Flore V et al. “Ultra-high
scale cytometry-based cellular interaction mapping”, <em>Nature
Methods</em> (2025)</li>
</ul>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<ul>
<li>Vonficht D, Jopp-Saile L, Yousefian S, Flore V et al. Ultra-high
scale cytometry-based cellular interaction mapping. <em>Nature
Methods</em> (2025).</li>
<li>Hao Y et al. Dictionary learning for integrative, multimodal and
scalable single-cell analysis. <em>Nat Biotechnol</em> 42, 293–304
(2024). <a href="https://doi.org/10.1038/s41587-023-01767-y" class="external-link uri">https://doi.org/10.1038/s41587-023-01767-y</a><br>
</li>
<li>Heumos L, Schaar A.C. et al. Best practices for single-cell analysis
across modalities. <em>Nat Rev Genet</em> (2023). <a href="https://doi.org/10.1038/s41576-023-00586-w" class="external-link uri">https://doi.org/10.1038/s41576-023-00586-w</a><br>
</li>
<li>Parks B &amp; Abdi I. BPCells: Single Cell Counts Matrices to PCA. R
package version 0.2.0 (2024) <a href="https://bnprks.github.io/BPCells" class="external-link uri">https://bnprks.github.io/BPCells</a>
</li>
</ul>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Lea Jopp-Saile, Viktoria Flore.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
